x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlConfigMigrator.java

```diff
@@ -0,0 +1,386 @@
+                ClusterState.Builder newState = ClusterState.builder(currentState);
+                newState.metaData(MetaData.builder(currentState.getMetaData())
+                        .putCustom(MlMetadata.TYPE, removed.mlMetadata)
+                        .build());
+                return newState.build();
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                if (removedConfigs.get() != null) {
+                    if (removedConfigs.get().removedJobIds.isEmpty() == false) {
+                        logger.info("ml job configurations migrated: {}", removedConfigs.get().removedJobIds);
+                    }
+                    if (removedConfigs.get().removedDatafeedIds.isEmpty() == false) {
+                        logger.info("ml datafeed configurations migrated: {}", removedConfigs.get().removedDatafeedIds);
+                    }
+                }
+                listener.onResponse(Boolean.TRUE);
+            }
+        });
+    }
+
+    static class RemovalResult {
+        MlMetadata mlMetadata;
+        List<String> removedJobIds;
+        List<String> removedDatafeedIds;
+
+        RemovalResult(MlMetadata mlMetadata, List<String> removedJobIds, List<String> removedDatafeedIds) {
+            this.mlMetadata = mlMetadata;
+            this.removedJobIds = removedJobIds;
+            this.removedDatafeedIds = removedDatafeedIds;
+        }
+    }
+
+    /**
+     * Remove the datafeeds and jobs listed in the parameters from
+     * mlMetadata if they exist. An account of emoved jobs and datafeeds
+     * is returned in the result structure alongside a new MlMetadata
+     * with the config removed.
+     *
+     * @param jobsToRemove       Jobs
+     * @param datafeedsToRemove  Datafeeds
+     * @param mlMetadata         MlMetadata
+     * @return Structure tracking which jobs and datafeeds were actually removed
+     * and the new MlMetadata
+     */
+    static RemovalResult removeJobsAndDatafeeds(List<String> jobsToRemove, List<String> datafeedsToRemove, MlMetadata mlMetadata) {
+        Map<String, Job> currentJobs = new HashMap<>(mlMetadata.getJobs());
+        List<String> removedJobIds = new ArrayList<>();
+        for (String jobId : jobsToRemove) {
+            if (currentJobs.remove(jobId) != null) {
+                removedJobIds.add(jobId);
+            }
+        }
+
+        SortedMap<String, DatafeedConfig> currentDatafeeds = new TreeMap<>(mlMetadata.getDatafeeds());
+        List<String> removedDatafeedIds = new ArrayList<>();
+        for (String datafeedId : datafeedsToRemove) {
+            if (currentDatafeeds.remove(datafeedId) != null) {
+                removedDatafeedIds.add(datafeedId);
+            }
+        }
```

Comment:
No this was because `MlMetaData.getDatafeeds` returns a sorted map whereas `MlMetaData.getJobs` returns a plain map and I brought that inconsistency into this code. I change it to a plain map as it does not need to be anything else.